File: src/schemas/prediction.py
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


class PredictionRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    age: int = Field(ge=0)
    sex: Literal["female", "male"]
    bmi: float = Field(gt=0, allow_inf_nan=False)
    children: int = Field(ge=0)
    smoker: Literal["yes", "no"]
    region: Literal["northeast", "northwest", "southeast", "southwest"]


class PredictionResponse(BaseModel):
    predicted_charges: float = Field(allow_inf_nan=False)
    currency: Literal["USD"] = "USD"


class HealthResponse(BaseModel):
    status: Literal["ok"] = "ok"

File: frontend/src/types/prediction.ts
export const SEX_VALUES = ['female', 'male'] as const
export const SMOKER_VALUES = ['yes', 'no'] as const
export const REGION_VALUES = [
  'northeast',
  'northwest',
  'southeast',
  'southwest',
] as const

export type Sex = (typeof SEX_VALUES)[number]
export type Smoker = (typeof SMOKER_VALUES)[number]
export type Region = (typeof REGION_VALUES)[number]

export interface PredictionRequest {
  age: number
  sex: Sex
  bmi: number
  children: number
  smoker: Smoker
  region: Region
}

export interface PredictionResponse {
  predicted_charges: number
  currency: string
}

export interface PredictionFormValues {
  age: string
  sex: string
  bmi: string
  children: string
  smoker: string
  region: string
}

export type PredictionField = keyof PredictionFormValues
export type FieldErrors = Partial<Record<PredictionField, string>>
